[ Frontend ] Multiprocessing for OpenAI Server with `zeromq`

robertgshaw2-redhat opened the pull request:
RFC: https://github.com/vllm-project/vllm/issues/6797

SUMMARY:
* Use multiprocessing for OpenAI Server such that it does not contend with `AsyncLLMEngine`
* Use `zeromq` for networking between OAI process and `AsyncLLMEngine`
* Currently uses pickling for python serialization, to update this to use protobufs in separate PR
* Gives ~20% TPOT improvement for QPS 10 for `llama-8b-instruct` on H100 (550 input tokens, 150 output tokens)

TODO:

- [x] enabling aborting
- [x] enable do_log_stats
- [x] error handling
- [x] refactor to use single socket
- [x] use open port
- [x] turn on the other OAI endpoints
- [x] add flag to disable mp
- [x] tokenizer 
- [x] guided decoding
- [x] tracing
- [x] make CI green

PERFORMANCE:
* `zeromq` based protocol is delivering speedup in high QPS scenario:
*  Llama-3-8B-Instruct, 1xH100, 550 input tokens | 150 output tokens, 10 QPS, chunked prefill

| Metric | `main` | `grpc` | `zmq + pickle` (this PR) | `zmq + protobuf` |
| - | - | - | - | - |
| TPOT (mean) | 70.74 | 64.15 | 57.35 | 54.54 |
| TTFT (mean) | 687.69 | 450.05 | 298.09 | 269.01 |


Clear win + clear we should move forward with `zeromq`

---

Benchmark script for reference:

```bash
MODEL="meta-llama/Meta-Llama-3-8B-Instruct"
TOTAL_SECONDS=60
QPS_RATES=("10")

for QPS in ${QPS_RATES[@]}; do
    NUM_PROMPTS=$((TOTAL_SECONDS * QPS))
    echo "===== RUNNING NUM_PROMPTS = $NUM_PROMPTS QPS = $QPS ====="

    python3 benchmarks/benchmark_serving.py \
        --model $MODEL \
        --dataset-name sonnet --sonnet-input-len 550 --sonnet-output-len 150 --dataset-path benchmarks/sonnet.txt \
        --num-prompts $NUM_PROMPTS --request-rate $QPS
done
```

--- 

FOLLOW UP WORK:
* Refactor how logits processors are handled. They should not belong to the API layer anymore, rather below the engine. Currently there are hacks around making it serializable. 
* Switch to using a `protobuf` communication layer for the RPC Server
* Move detokenization to the OAI layer
* Profile using `N` uvicorn processes + expose a flag
* Support embedding models via RPC

joerunde committed:
:alembic: add backend proto file

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:recycle: move proto to grpc/pb

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:sparkles: add proto compilation

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

Commit:
updated

Commit:
kinda working

joerunde committed:
:construction: more wip

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

Commit:
fixed

joerunde committed:
:bug: fixup race condition

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: remove timeout

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

Commit:
format

Commit:
streaming

Commit:
removed breaks

Commit:
pushing current state

joerunde committed:
:alembic: try unix sockets

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:zap: no background loop

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

Commit:
spurious change

Commit:
remove spurious change

Commit:
spurious changes

Commit:
spurioous change

joerunde committed:
:bug: whoops

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:memo: log stuff

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

Commit:
stash

Commit:
pushing up

Commit:
stash

Commit:
actually working

Commit:
cleanup

Commit:
more cleanup

Commit:
cleanup

Commit:
stash

Commit:
more cleanup

Commit:
setup

Commit:
cleanup

Commit:
format

Commit:
cleaning up

Commit:
zlib

Commit:
Revert "zlib"

This reverts commit e42be96fc9f03b09a89d19e0a57ee936bf8cdc3a.

Commit:
turn on chunked prefill

Commit:
move RPC code into oai server

Commit:
format

Commit:
format

Commit:
trying to flow it through

Commit:
cleaning

Commit:
cleaning

Commit:
cleaning

Commit:
add stubs

Commit:
format

Commit:
working with single launch...

Commit:
working end to end - with some hacks

mgoin commented on vllm/entrypoints/openai/rpc/__init__.py:
```diff
@@ -0,0 +1,26 @@
+from vllm.prompt_adapter.request import PromptAdapterRequest
+from vllm.sampling_params import SamplingParams
+
+VLLM_GENERATE_RPC_PATH = "tcp://localhost:5570"
+VLLM_GET_DATA_RPC_PATH = "tcp://localhost:5571"
+VLLM_IS_READY_RPC_PATH = "tcp://localhost:5572"
```
At the very least we should be choosing randomly available ports above the "user space" of the 1000s port range. See https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.bind_to_random_port

joerunde replied:
@mgoin yeah that'd be better. Any idea how we'd notify the clients what port to connect to in that case?

The server that would be calling `.bind_to_random_port()` is in a different process than the openai server that needs to connect clients to it

youkaichao replied:
the the openai server can listen first, pass the ports to the server process, and then the server just connects to it?

mgoin commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -0,0 +1,104 @@
+
+    async def run_loop(self):
+        # Notify the RPC client that we are ready to recieve requests.
+        await self.is_ready_socket.send_string("Ready!")
```
This ready string should be a global shared constant

mgoin commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -0,0 +1,104 @@
+                running_tasks.add(task)
+                task.add_done_callback(running_tasks.discard)
+
+        # TODO: Do I need to close the generate / get_data sockets?
```
I think it's best to close sockets explicitly if possible per https://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.close

> If this is not called, the socket will automatically be closed when it is garbage collected, in which case you may see a ResourceWarning about the unclosed socket.

joerunde committed:
:goal_net: handle shutdown and request errors

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

WoosukKwon commented:
Awesome! This is really great!

joerunde committed:
:art: fmt and clean up shutdown handler

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fixup type hint for queue

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:sparkles: update chat endpoint

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fixup zmq constant types

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:sparkles: hook up de/tokenize

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:recycle: add VLLMBackend protocol

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

njhill commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -62,12 +67,12 @@ async def lifespan(app: fastapi.FastAPI):
         while True:
             await asyncio.sleep(10)
-            await engine.do_log_stats()
+            await backend.do_log_stats()
```
I think this should just move into AsyncLLMEngine right? (perhaps whether it's started controlled by some arg)

robertgshaw2-neuralmagic replied:
Fine with me if it works

njhill commented on vllm/entrypoints/openai/serving_completion.py:
```diff
@@ -93,10 +93,10 @@ async def create_completion(self, request: CompletionRequest,
             ) = self._maybe_get_adapters(request)
 
-            tokenizer = await self.engine.get_tokenizer(lora_request)
+            tokenizer = await self.vllm_backend.get_tokenizer(lora_request)
```
Presumably the idea is that these would never actually be remote calls and instead live within the "client" layer of the remote VLLMBackend impl (get_tokenizer, get_decoding_config, is_tracing_enabled, etc).

robertgshaw2-neuralmagic replied:
We we should just pull over all the state we need at the start once

njhill replied:
Yes exactly, no problem with that being common logic with the single-process case too.

Though get_tokenizer might need to be slighlty different, since we may need to load lora-speciic tokenizers on the fly.

njhill commented on vllm/utils.py:
```diff
@@ -315,7 +322,13 @@ async def producer(i: int, iterator: AsyncIterator[T]):
     async def consumer():
         try:
             while not all(finished) or not queue.empty():
```
I think we can eliminate the `finished` list and just have a count that's decremented in the consumer when it receives each `ProducerFinished`?

njhill commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -0,0 +1,106 @@
+
+        # Send GenerateRequest to the RPC Server.
+        await socket.send_multipart([
+            pickle.dumps(
```
As @robertgshaw2-neuralmagic suggested, let's use msgspec?

robertgshaw2-neuralmagic replied:
Let's separate the messaging protocol optimizations to a separate PR

njhill reviewed (commented):
Thanks @joerunde @robertgshaw2-neuralmagic this looks great so far!

joerunde committed:
Frontend mp flag (#384)

@robertgshaw2-neuralmagic 

This adds the `--disable-frontend-multiprocessing` flag and should also
correctly pick up embeddings models to disable the multiprocessing here.
(Also some unrelated formatting changes)

The backend stuff is wrapped up in a context manager that handles the
process startup and shutdown at exit as well, so that we don't have to
muck around much in the existing server lifecycle code

---------

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

robertgshaw2-redhat committed:
Features / Cleanup for MP Frontend (#387)

SUMMARY:
* refactor to use single socket
* cleanup comments / logging
* add `do_log_stats`
* add `abort`

joerunde committed:
Use random port for backend (#390)

Picks an open port to use and boots both the client and server with it

---------

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

njhill committed:
Await socket operations + some other minor cleanup (#391)

joerunde committed:
:sparkles: health check round 2 (#392)

With all the extra fun refactors

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

robertgshaw2-redhat committed:
Add tokenizer (#394)

SUMMARY:
* add endpoints to request `ModelConfig`, `SchedulerConfig`,
`LoRAConfig`, `ParallelConfig`
* factor out tokenizer group creation function to be a utility function
* create tokenizer_group on client side

joerunde committed:
Socket context (#393)

Ensures no sockets are leaked on the client-side

Also postpones the server shutdown await so that the backend can
shutdown concurrently, and all connections can be cleaned up at the same
time. This prevents hangs where the frontend blocks on remaining
connections but the backend has not yet initiated shutdown

---------

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

robertgshaw2-redhat committed:
Logit bias (#395)

SUMMARY:
* fix issue with logit bias loading

joerunde committed:
Merge remote-tracking branch 'upstream/main' into isolate-oai-server-process

joerunde committed:
:bug: messed up the revert in the merge commit :(

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

robertgshaw2-redhat committed:
fix (#396)

SUMMARY:
* passed clamped

joerunde committed:
Merge remote-tracking branch 'upstream/main' into isolate-oai-server-process

Commit:
format

Commit:
stash

robertgshaw2-redhat committed:
Fix failed tests (#398)

SUMMARY:
* hack

Commit:
Merge branch 'main' into isolate-oai-server-process

Commit:
fixed merge conflicts

Commit:
updated

robertgshaw2-neuralmagic commented:
/ready

Commit:
cleaning

joerunde committed:
:white_check_mark: add test for multiprocessing flag (#399)

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:sparkles: pipe tracing flag (#400)

(plus rounding out the protocol with an error on `.encode`)

---------

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

simon-mo commented on vllm/engine/protocol.py:
```diff
@@ -0,0 +1,84 @@
+
+
+@runtime_checkable
+class VLLMBackend(Protocol):
```
Naming wise, I think this the protocol for `AsyncEngineClient` or `AsyncEngine`. Also I prefer no vLLM in the class name. The entire package is named under vLLM. Backend is ambiguous because we already use it for hardware backends, and attention backends.

joerunde replied:
Yeah `AsyncEngine` is pretty spot on

robertgshaw2-neuralmagic replied:
I think we should call it a client

joerunde replied:
I don't think we should call it a client, this is the name of the interface that both the `AsyncLLMEngine` itself and the client to the backend process running the `AsyncLLMEngine` implement.

I'd be up for renaming like `RPCClient` -> `AsyncLLMEngineClient` and `RPCServer` -> `AsyncLLMEngineServer` if we wanna be consistent with these names

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -72,6 +77,60 @@ async def _force_log():
+    global backend
+
+    # First need to determine if this is an embeddings model
+    # (no remote backend for those)
```
why?

joerunde replied:
cc @robertgshaw2-neuralmagic 

I think because it'd be more work to hook those up, and we're not worried about the throughput improvements for embeddings atm

robertgshaw2-neuralmagic replied:
Planning to do follow to enable RPC for Embedding

simon-mo replied:
leave a comment saying future work plz.

simon-mo commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -72,6 +77,60 @@ async def _force_log():
+
+    # First need to determine if this is an embeddings model
+    # (no remote backend for those)
+    model_config = ModelConfig(model=args.model,
+                               tokenizer=args.tokenizer,
+                               tokenizer_mode="auto",
+                               trust_remote_code=False,
+                               seed=0,
+                               dtype="float16")
+    if model_config.embedding_mode or args.disable_frontend_multiprocessing:
```
make this a util? 
`def model_is_embedding(model, tokenizer)`

Commit:
integration tests for old backend

simon-mo reviewed (commented):
reviewed non `rpc` code. recommend the rename from vllm_backend back to engine or engine client.

simon-mo commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -0,0 +1,216 @@
+import signal
+from typing import Any, Coroutine
+
+import cloudpickle
```
need this in requirements txt

robertgshaw2-neuralmagic replied:
I can explictily add it, but it seems to be installed already

simon-mo commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -0,0 +1,216 @@
+from vllm.logger import init_logger
+from vllm.usage.usage_lib import UsageContext
+
+logger = init_logger('vllm.entrypoints.openai.rpc.server')
```
```suggestion
logger = init_logger(__name__)
```

simon-mo reviewed (commented):
I think any non generation request can be serialized by `pickle` instead of `cloudpickle`, which is preferred performance wise, and prevent accidentally leaking large data structure.

simon-mo reviewed (commented):
LGTM over all. This is a good direction. Can we check one edge case? 
* Send request generate 2K toks
* Drop the connection 
* Check the `generate` coroutine is properly cancelled, and the abort it called, and there is no leakage of coroutines in both client and server (by printing out something...)

Commit:
rename

Commit:
cleaning

robertgshaw2-neuralmagic commented on vllm/engine/protocol.py:
```diff
@@ -0,0 +1,84 @@
+    ) -> PreTrainedTokenizer:
+        """Get the appropriate Tokenizer for the request"""
+
+    async def is_tracing_enabled(self) -> bool:
```
@joerunde why are these `pass`?

Commit:
ordering

Commit:
fix embedding model feedback

robertgshaw2-redhat committed:
Update vllm/entrypoints/openai/rpc/server.py

Co-authored-by: Simon Mo <simon.mo@hey.com>

robertgshaw2-neuralmagic commented:
> LGTM over all. This is a good direction. Can we check one edge case?
> 
> * Send request generate 2K toks
> * Drop the connection
> * Check the `generate` coroutine is properly cancelled, and the abort it called, and there is no leakage of coroutines in both client and server (by printing out something...)

Do you want this in automation?

Commit:
format

Commit:
Merge branch 'main' into isolate-oai-server-process

robertgshaw2-neuralmagic commented:
| Branch | 2 |  6 | 10 | 14 |
| - | - | - | - | - |

simon-mo reviewed (approved):
:shipit:

youkaichao commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -0,0 +1,248 @@
+    async def _send_one_way_rpc_request(self, request: RPC_REQUEST_TYPE,
+                                        error_message: str):
+        """Send one-way RPC request to trigger an action."""
+        with self.socket() as socket:
```
I'm trying to understand this. We create a socket every time we call this function?

robertgshaw2-neuralmagic replied:
Yes. I believe this is a common paradigm for zeromq based clients

also, this function is mostly used in setup

robertgshaw2-neuralmagic replied:
Also, this does not mean a new Unix socket is created. This is a zeromq socket. how zeromq handles this internally is opaque to us

youkaichao commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -72,6 +86,52 @@ async def _force_log():
+    # Otherwise, use the multiprocessing AsyncLLMEngine.
+    else:
+        # Start RPCServer in separate process (holds the AsyncLLMEngine).
+        port = get_open_port(envs.VLLM_RPC_PORT)
+        rpc_server_process = Process(target=run_rpc_server,
+                                     args=(engine_args,
+                                           UsageContext.OPENAI_API_SERVER,
+                                           port))
```
why do we need a new env var here? isn't `get_open_port()` enough?

robertgshaw2-neuralmagic replied:
I would be open to reverting this. Will just need to check how it interacts with tp

youkaichao reviewed (commented):
Thanks for the great work! I just left some nit comments that I don't understand. If you have time, could you please briefly draw some diagrams to show how many processes do we have use and how do they interact with each other?

robertgshaw2-neuralmagic commented on tests/entrypoints/openai/test_disable_mp.py:
```diff
@@ -0,0 +1,715 @@
+"""
+Repeat of tests in test_completion.py with the non-mp backend.
```
Because we need to modify default server args which is a fixture for the whole file so unfortunately I think this has to be a separate file
